File: Sort_Algo/qkSort.cpp
#include<iostream>
using namespace std;
 
int part(int arr[],int low,int high){
int pv = arr[low];
int i=low+1;
int j=high;
int temp;

do{
    while(arr[i]<pv){
        i++;
    }
    while(arr[j]>pv){
        j--;
    }
    if(i<j){
        temp=arr[i];
        arr[i]=arr[j];
        arr[j]=temp;
    }

}while(i<j);
temp=arr[low];
arr[low]=arr[j];
arr[j]=temp;

return j;
}
void qkSort(int arr[],int low,int high){
    int pv;
    if(low<high){
     pv = part(arr,low,high);
     qkSort(arr,low,pv-1);
     qkSort(arr,pv+1,high);
    }
}

int main(){
    int arr[]={4,3,2,1,16,8,7,29,5};
    int size = sizeof(arr)/sizeof(arr[0]);

    qkSort(arr,0,size-1);

    for(int i=0;i<size;i++){
        cout<<arr[i]<<" ";
    }
    return 0;
}

File: JAVASCRIPT/SORT_ALGO/qksort.js
function part(arr, low, high) {
    let pv = arr[low];
    let i = low + 1;
    let j = high;
    let temp;

    do {
        while (i <= high && arr[i] < pv) {
            i++;
        }
        while (arr[j] > pv) {
            j--;
        }
        if (i < j) {
            temp = arr[j];
            arr[j] = arr[i];
            arr[i] = temp;
        }
    } while (i < j);

    // Swap pivot with arr[j]
    temp = arr[j];
    arr[j] = arr[low];
    arr[low] = temp;
    
    return j;
}

function qksort(arr, low, high) {
    if (low < high) {
        let pv = part(arr, low, high);
        qksort(arr, low, pv - 1);
        qksort(arr, pv + 1, high);
    }
}

function main() {
    const arr = [5, 4, 6, 3, 7, 2, 8];
    const s = arr.length;

    qksort(arr, 0, s - 1);  // Note: s-1 here
    
    // Print the sorted array
    for (let i = 0; i < arr.length; i++) {
        console.log(arr[i]);
    }
}

main();